josh-mysql/src/index.js

```diff
@@ -0,0 +1,556 @@
+const mysql = require("mysql");
+
+// Lodash should probably be a core lib but hey, it's useful!
+const {
+  get: _get,
+  set: _set,
+  isNil,
+  isArray,
+  isFunction,
+  flatten,
+  cloneDeep,
+  unset,
+} = require('lodash');
+
+
+// Custom error codes with stack support.
+const Err = require('./error.js');
+// , buildQuery
+const {
+  getPaths,
+  serializeData,
+  // serialize,
+  // isObject,
+  // onChange,
+} = require('./utils.js');
+// const db = require("../../test.js");
+
+module.exports = class JoshProvider {
+  constructor(options) {
+    if (options.inMemory) {
+      throw new Error("Josh-MySQL doesn't support in-memory databases")
+    } else {
+      if (!options.name) throw new Error('Must provide options.name');
+
+      this.name = options.name;
+      this.validateName();
+      this.db = options.connection;
+
+      // find a way to improve that
+      if (!this.db || this.db.constructor.name !== "Connection") throw new Error("Must provider options.connection");
+    }
+  }
+
+  /**
+   * Internal method that translates a callback-based query into a promise-based query
+   * @private
+   * @param {mysql.QueryOptions | string} options The query
+   * @param {*} values Auto-completion values
+   * @param {boolean} [many=false] Whether to return multiple resulsts 
+   * @returns {Promise<*>} Results
+   */
+  query(options, values = [], many = false) {
+    return new Promise((resolve, reject) => {
+      this.db.query(options, values, (err, results, fields) => {
+        if (err) return reject(err);
+        resolve(many ? results : results[0]);
+      })
+    })
+  }
+
+  /**
+   * Executes transactions
+   * @param {[sql: string, values: string[]][]} transactions The transactions
+   * @returns {Promise<void | Error>} Returns `void` when everything is ok
+   */
+  runMany(transactions) {
+    return new Promise((resolve, reject) => {
+
+      this.db.beginTransaction(async err => {
+        if (err) return reject(err);
+
+        try {
+
+          for (let [sql, values] of transactions) {
+            await this.query(sql, values);
+          }
+          this.db.commit(resolve);
+
+        } catch (e) {
+          this.db.rollback(() => reject(e));
+        }
+      });
+
+    });
+
+  }
+
+  /**
+   * Internal method called on persistent Josh to load data from the underlying database.
+   * @param {Map} Josh In order to set data to the Josh, one must be provided.
+   * @returns {Promise} Returns the defer promise to await the ready state.
+   */
+  async init() {
+    // Test request
+    await this.query("SELECT 1;");
+
+    const database = await this.query("SELECT DATABASE();")
+      .catch(console.error);
+
+    if (!database || !database["DATABASE()"]) throw new Error("Connection must have a database");
+
+    await this.query(`CREATE TABLE IF NOT EXISTS ${this.name} (objkey VARCHAR(255), path VARCHAR(255), val TEXT, PRIMARY KEY(objkey, path)) ENGINE=INNODB;`)
```

Comment:
I remember i tried to to it, but i believe that primary keys don't support the next text type

Reply:
Some mysql types :
```
VARCHAR (max 255 bytes) : for me it's the biggest text type that can be in a primary key
TINYTEXT (255 B)
TEXT (64 KB)
MEDIUMTEXT (16 MB)
LONGTEXT (4 GB)
```

But for value it would be a good idea to increase the size of data

Reply:
To compare, i'm reading that SQLite TEXT type can contain up to 10^9 bytes (1 GB if i'm good at math)

Reply:
See https://stackoverflow.com/questions/21700716/creating-mysql-a-table-with-a-primary-text-key

Reply:
Increased value size in https://github.com/eslachance/josh-providers/pull/17/commits/f49c51b493a795a87f0a76950dd83ebf9001128f